birdset/modules/models/ssast.py

```diff
@@ -0,0 +1,462 @@
+            # print(self.mask_correct)
+            # print(torch.eq(torch.argmax(self.softmax(total), dim=0), self.mask_correct))
+            # print([float(t)*99 for t in torch.eq(torch.argmax(self.softmax(total), dim=0), self.mask_correct)])
+
+            fold = torch.nn.Fold(output_size=([self.input_fdim, self.input_tdim]), kernel_size=(self.fshape, self.tshape), stride=(self.fstride, self.tstride))
+            pred = fold(pred.transpose(1, 2))
+            masked = fold(masked.transpose(1, 2))
+
+            return pred, masked
+
+    # # masked patch pretraining with generative objective
+    def mpg(self, input, mask_patch, cluster):
+        B = input.shape[0]
+        x = self.v.patch_embed(input)
+        input = self.unfold(input).transpose(1, 2)
+
+        # size 12(batch_size) * 100(#mask_patch), index of masked patches
+        mask_index = torch.empty((B, mask_patch), device=x.device, requires_grad=False).long()
+        # size 12(batch_size) * 512(sequence_len) * 768(hidden_dim)
+        mask_dense = torch.ones([x.shape[0], x.shape[1], x.shape[2]], device=x.device)
+        for i in range(B):
+            # randomly generate #mask_patch mask indexes without duplicate
+            if cluster == True:
+                # use this if you are masking e.g. 16*16 patches
+                mask_index[i] = self.gen_maskid_patch(self.num_patches, mask_patch)
+            else:
+                # use this if you are masking frame, i.e., 128*2 patches
+                mask_index[i] = self.gen_maskid_frame(self.num_patches, mask_patch)
+            mask_dense[i, mask_index[i], :] = 0
+
+        mask_tokens = self.mask_embed.expand(B, x.shape[1], -1)
+
+        # follow BEIT paper, mask with learnable masking embedding, but no performance diff observed compared with masking with 0s.
+        x = x * mask_dense + (1-mask_dense) * mask_tokens
+
+        # go through the Transformer layers
+        cls_tokens = self.v.cls_token.expand(B, -1, -1)  # stole cls_tokens impl from Phil Wang, thanks
+        dist_token = self.v.dist_token.expand(B, -1, -1)
+        x = torch.cat((cls_tokens, dist_token, x), dim=1)
+        x = x + self.v.pos_embed
+        x = self.v.pos_drop(x)
+        for blk in self.v.blocks:
+            x = blk(x)
+        x = self.v.norm(x)
+
+        pred = torch.empty((B, mask_patch, self.fshape * self.tshape), device=x.device).float()  # e.g. size 12*100*256
+        target = torch.empty((B, mask_patch, self.fshape * self.tshape), device=x.device).float() # e.g. size 12*100*256
+
+        for i in range(B):
+            #  +2 for indexes because cls and dis token
+            pred[i] = self.gpredlayer(x[i, mask_index[i] + self.cls_token_num, :])
+            target[i] = input[i, mask_index[i], :]
+
+        # calculate the MSE loss
+        mse = torch.mean((pred - target) ** 2)
+
+        return mse
+
+    def forward(self, input_values, task='ft_avgtok', cluster=True, mask_patch=400, labels=None):
```

Comment:
could you use a String Literal to get type safety?

Reply:
Forward function is much simpler now